File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) 
    {
        int left = 0,right = height.size()-1;
        int ans = 0;
        int leftMax = 0,rightMax = 0;
        while(left<right)
        {
            if(height[left] <= height[right])
            {
                if(height[left] >= leftMax)
                    leftMax = height[left];
                else
                ans+= abs(leftMax - height[left]);
                left++;
            }
            else
            {
                if(height[right] >= rightMax)
                    rightMax = height[right];
                else
                ans += abs(rightMax - height[right]);
                right--;
            }
        }
        return ans;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution(object):
    def trap(self, height):
        left , right = 0 , len(height) - 1
        leftMax , rightMax = 0,0
        ans = 0
        while left <= right:

            if height[left] <= height[right]:

                if leftMax > height[left]:
                    ans+=(leftMax - height[left])
                else:
                    leftMax = height[left]
                left+=1
            else:
                if rightMax > height[right]:
                    ans+= (rightMax - height[right])
                else:
                    rightMax = height[right]
                right-=1
        return ans

